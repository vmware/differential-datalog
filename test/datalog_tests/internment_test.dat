start;

insert internment_test.IInternedString("foo"),
insert internment_test.IInternedString("bar"),

commit dump_changes;

start;

insert internment_test.IStruct(internment_test.Tag1{true}, (-5, 5.5), 10000000000000000000000000),
insert internment_test.IStruct(internment_test.Tag2{10, "foo"}, (5_000, -1.2345), -10000000000000000000000000),

commit dump_changes;
