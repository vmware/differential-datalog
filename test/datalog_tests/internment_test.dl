input relation IInternedString(ix: istring)
relation StaticInternedString(ix: istring)

StaticInternedString(intern("static foo")).
StaticInternedString(i"ifoo").
StaticInternedString(i[|ibar|]).
StaticInternedString(i"ifoo${25}").
StaticInternedString(i"ifoo${25}" "!").
StaticInternedString(i$[|ibar${2+2}|]).

relation AllInternedString(ix: istring)
AllInternedString(ix) :- IInternedString(ix).
AllInternedString(ix) :- StaticInternedString(ix).

output relation OInternedString(x: string, ix: istring)

OInternedString(ival(s), s) :-
    AllInternedString(s).

OInternedString(ival(s1) ++ " " ++ ival(s2), intern(ival(s1) ++ " " ++ ival(s2))) :-
    AllInternedString(s1),
    AllInternedString(s2).
