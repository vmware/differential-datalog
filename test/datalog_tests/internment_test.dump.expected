internment_test.OInternedString:
internment_test.OInternedString{.x = "bar", .ix = "bar"}: +1
internment_test.OInternedString{.x = "bar bar", .ix = "bar bar"}: +1
internment_test.OInternedString{.x = "bar foo", .ix = "bar foo"}: +1
internment_test.OInternedString{.x = "bar ibar", .ix = "bar ibar"}: +1
internment_test.OInternedString{.x = "bar ibar4", .ix = "bar ibar4"}: +1
internment_test.OInternedString{.x = "bar ifoo", .ix = "bar ifoo"}: +1
internment_test.OInternedString{.x = "bar ifoo25", .ix = "bar ifoo25"}: +1
internment_test.OInternedString{.x = "bar ifoo25!", .ix = "bar ifoo25!"}: +1
internment_test.OInternedString{.x = "bar static foo", .ix = "bar static foo"}: +1
internment_test.OInternedString{.x = "foo", .ix = "foo"}: +1
internment_test.OInternedString{.x = "foo bar", .ix = "foo bar"}: +1
internment_test.OInternedString{.x = "foo foo", .ix = "foo foo"}: +1
internment_test.OInternedString{.x = "foo ibar", .ix = "foo ibar"}: +1
internment_test.OInternedString{.x = "foo ibar4", .ix = "foo ibar4"}: +1
internment_test.OInternedString{.x = "foo ifoo", .ix = "foo ifoo"}: +1
internment_test.OInternedString{.x = "foo ifoo25", .ix = "foo ifoo25"}: +1
internment_test.OInternedString{.x = "foo ifoo25!", .ix = "foo ifoo25!"}: +1
internment_test.OInternedString{.x = "foo static foo", .ix = "foo static foo"}: +1
internment_test.OInternedString{.x = "ibar bar", .ix = "ibar bar"}: +1
internment_test.OInternedString{.x = "ibar foo", .ix = "ibar foo"}: +1
internment_test.OInternedString{.x = "ibar4 bar", .ix = "ibar4 bar"}: +1
internment_test.OInternedString{.x = "ibar4 foo", .ix = "ibar4 foo"}: +1
internment_test.OInternedString{.x = "ifoo bar", .ix = "ifoo bar"}: +1
internment_test.OInternedString{.x = "ifoo foo", .ix = "ifoo foo"}: +1
internment_test.OInternedString{.x = "ifoo25 bar", .ix = "ifoo25 bar"}: +1
internment_test.OInternedString{.x = "ifoo25 foo", .ix = "ifoo25 foo"}: +1
internment_test.OInternedString{.x = "ifoo25! bar", .ix = "ifoo25! bar"}: +1
internment_test.OInternedString{.x = "ifoo25! foo", .ix = "ifoo25! foo"}: +1
internment_test.OInternedString{.x = "static foo bar", .ix = "static foo bar"}: +1
internment_test.OInternedString{.x = "static foo foo", .ix = "static foo foo"}: +1
